cli/src/commands/util.rs

```diff
@@ -109,13 +106,35 @@ fn cmd_util_completion(
     args: &UtilCompletionArgs,
 ) -> Result<(), CommandError> {
     let mut app = command.app().clone();
+    let warn = |shell| {
+        writeln!(
+            ui.warning(),
+            "warning: `jj util completion --{shell}` will be removed in a future version, and \
+             this will be a hard error"
+        )
+    };
     let mut buf = vec![];
-    let shell = if args.zsh {
-        clap_complete::Shell::Zsh
-    } else if args.fish {
-        clap_complete::Shell::Fish
-    } else {
-        clap_complete::Shell::Bash
+    let shell = match (args.shell, args.fish, args.zsh, args.bash) {
+        (Some(s), false, false, false) => s,
+        // allow `--fish` and `--zsh` for back-compat, but don't allow them to be combined
+        (None, true, false, false) => {
+            warn("fish")?;
+            Shell::Fish
+        }
+        (None, false, true, false) => {
+            warn("zsh")?;
+            Shell::Zsh
+        }
+        // default to bash for back-compat. TODO: consider making `shell` a required argument
+        (None, false, false, _) => {
+            warn("bash")?;
+            Shell::Bash
+        }
+        _ => {
+            return Err(user_error(
+                "cannot generate completion for multiple shells at once",
+            ))
```

Comment:
FYI, clap can check the conflict for us, like we do e.g. here: https://github.com/martinvonz/jj/blob/0f2715203f8272a2740beee5d583588cf48c4ff5/cli/src/commands/rebase.rs#L121

(But this is fine, especially since it's a temporary thing.)